### PR TITLE
Raise number of threads used by GASNet by default + add warning if exceeded

### LINKED ISSUE #26379: GASNet's max threads per process silently limiting Chapel's maxTaskPar
A user running on a high core-count node was surprised to find that they only got a value of `here.maxTaskPar` of 255 despite having more cores than that.  Digging into the source, they found that it was due to the runtime [deferring to the comm layer](https://github.com/chapel-lang/chapel/blob/main/runtime/src/tasks/qthreads/tasks-qthreads.c#L780-L784) (in this case, GASNet) in terms of determining how many threads to use.  After rebuilding with `CHPL_GASNET_CFG_OPTIONS="--with-max-pthreads-per-node=2000"` (a generous upper-bound), they were able to use all the cores on the node.

This issue is filed to ask what we could do about this.
* at a minimum, it seems that we should warn when this happens, with notes or pointers to documentation about how to increase the limit
* there was also a question from the user about whether this default should be raised in GASNet (or Chapel's configuration of GASNet) given modern processor design and core count increases
* maybe there's even something more significant that could be done to reduce the need for a configuration-time limit?

### PR DESCRIPTION
**Pull request by bradcray**
[co-developed by @bonachea]

As noted in https://github.com/chapel-lang/chapel/issues/26379, we've recently had a user on a high-core-count system get surprised when we silently dialed the maximum number of tasks down to 255 rather than using the number of cores on that system.  This PR addresses that issue in two ways:

* by inserting a warning if/when this happens so that the user can be aware that it is happening.  This warning is specialized to tell how to raise the limit for GASNet users (the only case that imposes a limit at present, we think).
* by raising the number of threads that GASNet is configured with to 65535 as suggested by @bonachea in https://github.com/chapel-lang/chapel/issues/26379#issuecomment-2529688584, which will result in 1024 being used as the default limit unless overridden with `GASNET_MAX_THREADS`

In testing this, I also updated the `chpl_comm_getMaxThreads()` call to use the soft limit from GASNet on the number of threads rather than the hard, configuration-time limit that we were using previously.  Since these two values have traditionally been the same (by default at least, assuming the user didn't override either), this has not been an issue, but now that we're configuring to one value and relying on GASNet's default for the other, it becomes more iimportant.  Without this change, the implication is that if the user had used more than 1024 threads, they could exceed the size of the arrays that GASNet had allocated and bad things would happen.

While here, I also noticed that in our Makefiles we seemed to be passing the list of GASNet configuration options to it twice, introduced in 2014 in d76598295b5.  I don't see any indication that that's anything more intentional than a typo, so here, I've removed one of the occurrences, leaving the one at the end of the line there, thinking it'd support overriding other, earlier options.  After the fact, Dan pointed out that he'd made the same change in #25414 — a PR that seems to have fallen between the cracks at some point.  I also adopted the `FORCE` is a phony target rule from there to permit it to be closed and opened https://github.com/chapel-lang/chapel/issues/26548 to note that we should do the same in other Makefiles.

Resolves #26379
Resolves #25414

### CODE REVIEW

File: runtime/src/tasks/qthreads/tasks-qthreads.c [hunk 1]
```diff
@@ -502,7 +502,17 @@ static void setupAvailableParallelism(int32_t maxThreads) {
...
-            hwpar = maxThreads;
+          if (chpl_nodeID == 0) {
+            char msg[1024];
+            snprintf(msg, sizeof(msg),
+                     "The CHPL_COMM setting is limiting the number of threads "
+                     "to %d, rather than the hardware's preference of %d.  "
+                     "If using CHPL_COMM=gasnet, set 'GASNET_MAX_THREADS' to "
+                     "a higher number in your environment to address this.\n",
+                     maxThreads, hwpar);
+            chpl_warning(msg, 0, 0);
```
Comment (bonachea) [inline runtime/src/tasks/qthreads/tasks-qthreads.c:513]: New warning text LGTM, and should usually achieve the goal of avoiding silent under-utilization of cores.

However IIUC the new warning won't be printed on a heterogeneous system where node 0 has a smaller core count than `maxThreads`, but some other node might have a higher core count.

Not sure if such node heterogeneity is of interest (or even permitted by Chapel?) so feel free to ignore this observation if it seems irrelevant or not worth addressing.

File: third-party/gasnet/Makefile [hunk 2]
```diff
@@ -135,7 +135,7 @@ gasnet-config: FORCE
 	mkdir -p $(GASNET_BUILD_DIR)
 	cd $(GASNET_SUBDIR) && ./Bootstrap -T
 	$(XTRA_CONFIG_COMMAND)
-	$(ENSURE_GASNET_COMPATIBLE_COMPILER) cd $(GASNET_BUILD_DIR) && $(CHPL_GASNET_ENV_VARS) $(GASNET_SUBDIR)/$(CHPL_GASNET_CFG_SCRIPT) --prefix=$(GASNET_INSTALL_DIR) $(CHPL_GASNET_CFG_OPTIONS) --disable-seq --enable-par --disable-parsync $(CHPL_GASNET_CFG_OPTIONS)
+	$(ENSURE_GASNET_COMPATIBLE_COMPILER) cd $(GASNET_BUILD_DIR) && $(CHPL_GASNET_ENV_VARS) $(GASNET_SUBDIR)/$(CHPL_GASNET_CFG_SCRIPT) --prefix=$(GASNET_INSTALL_DIR) --disable-seq --enable-par --disable-parsync --with-max-pthreads-per-node=65535 $(CHPL_GASNET_CFG_OPTIONS)
```
Comment (bonachea) [inline third-party/gasnet/Makefile:138]: This hunk partially overlaps the same fix proposed in the first commit of #25414 for the same cosmetic defect.

It sounds like we both independently made the same observation about the GASNet configure line and arrived at the same fix, so feel free to close #25414 in favor of this PR.

**Review by bonachea — COMMENTED** [review]
Thanks for the fixes!

Made a few minor observations, in case they're helpful.

File: runtime/src/tasks/qthreads/tasks-qthreads.c (hunk 1 above)
Comment (bradcray) [reply runtime/src/tasks/qthreads/tasks-qthreads.c:513]: @bonachea : I'm finally finding my way back to this PR.  Thanks for the quick look and comments, and sorry not to have acknowledged them before now.

Chapel doesn't worry overmuch about node heterogenity [anymore], primarily because peoples' questions about it seem to typically be academic.  That said, I'm also not interested in making things worse/harder for node heterogeneity than they are, so am still weighing the question seriously.

I'm 96% certain that this runtime startup code is executed in an SPMD manner (one process per locale), so would expect this logic to fire on every locale, and therefore to potentially take different actions on different locales.  To check that theory, I put a printf() into the code here, ran a hello world on 4 locales, and got the message printed 4 times.  Does that reassure you, or are you seeing a subtlety that I'm overlooking in that analysis?

File: third-party/gasnet/Makefile (hunk 2 above)
Comment (bradcray) [reply third-party/gasnet/Makefile:138]: Sheesh I'm embarrassed about how long that PR has been sitting there.  I may have mentioned this before, but we're generally not as good about monitoring new PRs unless they're pointed out to us as we are issues and questions in other forums, so definitely feel free to ping someone / us / me on a PR to get attention to it.  Or maybe you did that off the PR and someone just dropped the ball…  Anyway, thanks for pointing that out.

Comment (bonachea) [reply third-party/gasnet/Makefile:138]: Feel free to review/merge #25414 yourself now that it's been pointed out ;-)

File: runtime/src/tasks/qthreads/tasks-qthreads.c (hunk 1 above)
Comment (bonachea) [reply runtime/src/tasks/qthreads/tasks-qthreads.c:513]: @bradcray Thanks for the context. 

> expect this logic to fire on every locale, and therefore to potentially take different actions on different locales. To check that theory, I put a printf() into the code here, ran a hello world on 4 locales, and got the message printed 4 times. Does that reassure you, or are you seeing a subtlety that I'm overlooking in that analysis?

FWIW the logic I was specifically referring to is the invocation of `chpl_warning()` on line 513 which is conditional on line 505: `if (chpl_nodeID == 0) {` .  Are you saying that `chpl_nodeID` is zero on all locales? 

As I said in my initial comment, I'm fine with dropping this observation entirely in light of Chapel's position on heterogeneous systems. I was really just noting it for completeness.

Comment (bradcray) [reply runtime/src/tasks/qthreads/tasks-qthreads.c:513]: [of course, the downside to my statement is that we'll probably get a message per locale on homogeneous systems where we hit this, which may be overkill and regrettable, but still feels like an improvement over the status quo]

Comment (bradcray) [reply runtime/src/tasks/qthreads/tasks-qthreads.c:513]: [whoops, made the above comment without seeing your earlier response.  I think I must've been in an un-refreshed browser tab]

> FWIW the logic I was specifically referring to is the invocation of chpl_warning() on line 513 which is conditional on line 505: if (chpl_nodeID == 0) { . Are you saying that chpl_nodeID is zero on all locales?

Oh, duh… 🤦  And of course I put my printf outside of that conditional rather than within it.  :P

I think I'm inclined to proceed with the PR as-is for simplicity and due to not really being aware of realistic heterogeneous use cases (combined with a hope that, with the new setting, we'll hit this conditional far less often than we would've had it been here before, with the configuration changes).

File: third-party/gasnet/Makefile (hunk 2 above)
Comment (bradcray) [reply third-party/gasnet/Makefile:138]: Before seeing this, I pulled your other change from that into this PR and added contributor credit for you to this one, mostly based on your previous comment and because it seemed simpler.

File: runtime/src/tasks/qthreads/tasks-qthreads.c (hunk 1 above)
Comment (bonachea) [reply runtime/src/tasks/qthreads/tasks-qthreads.c:513]: I share your hope and agree this seems like a reasonable resolution.

File: third-party/gasnet/Makefile (hunk 2 above)
Comment (bonachea) [reply third-party/gasnet/Makefile:138]: Works for me, I'll close the other PR.

File: runtime/src/tasks/qthreads/tasks-qthreads.c [hunk 3]
```diff
@@ -502,7 +502,17 @@ static void setupAvailableParallelism(int32_t maxThreads) {
     if (hwpar > 0) {
         // Limit the parallelism to the maximum imposed by the comm layer.
         if (0 < maxThreads && maxThreads < hwpar) {
-            hwpar = maxThreads;
+          if (chpl_nodeID == 0) {
+            char msg[1024];
+            snprintf(msg, sizeof(msg),
+                     "The CHPL_COMM setting is limiting the number of threads "
+                     "to %d, rather than the hardware's preference of %d.  "
+                     "If using CHPL_COMM=gasnet, set 'GASNET_MAX_THREADS' to "
```
Comment (jabraham17) [inline runtime/src/tasks/qthreads/tasks-qthreads.c:510]: I think this would be better if it just check against `CHPL_COMM` to specialize the error message for GASNet. But my understanding of this PR is that in practice this warning can only occur with GASNet, so I say leave it unless we see it occur on non-gasnet systems

**Review by jabraham17 — APPROVED** [review]

File: third-party/gasnet/Makefile [hunk 4]
```diff
@@ -158,4 +158,6 @@ gasnet: gasnet-config gasnet-build gasnet-install
 
 FORCE:
 
+.PHONY: FORCE
+
```
Comment (bonachea) [inline third-party/gasnet/Makefile:162]: I should note in passing that the same defect repaired by the `.PHONY` directive above for third-party/gasnet/Makefile alone also impacts many other Makefiles in the Chapel repo. I just didn't feel I had the authority to propose adding this to all of them ;-)

File: runtime/src/tasks/qthreads/tasks-qthreads.c (hunk 3 above)
Comment (bonachea) [reply runtime/src/tasks/qthreads/tasks-qthreads.c:510]: >  in practice this warning can only occur with GASNet

I believe that's accurate for today's backends.

**Review by bonachea — APPROVED** [review]

File: third-party/gasnet/Makefile (hunk 4 above)
Comment (bradcray) [reply third-party/gasnet/Makefile:162]: Yeah, I've felt long aware of that as well.  Though I'm also not remembering the consequences of not listing it in a `.PHONY` (and the fact that it hasn't seemed to bite us has prevented me from doing anything to improve the situation).  Can you remind me?

File: runtime/src/tasks/qthreads/tasks-qthreads.c (hunk 3 above)
Comment (bradcray) [reply runtime/src/tasks/qthreads/tasks-qthreads.c:510]: > I think this would be better if it just checks against CHPL_COMM

I agree that'd be better.  In putting this together very rapidly, I couldn't remember how to check `CHPL_COMM` within the runtime, nor how easy it was, and was apparently too lazy to look.  Now that Jade's reminded me, I'm going to tighten this up, less because I think it'll ever happen in practice (well, in the near-term anyway) and more because I don't really like the "If…" conditional in the warning when we really ought to know the answer.

File: third-party/gasnet/Makefile (hunk 4 above)
Comment (bonachea) [reply third-party/gasnet/Makefile:162]: Here is the [GNU documentation for PHONY targets](https://www.gnu.org/software/make/manual/html_node/Phony-Targets.html).

The intent of the `FORCE` target used throughout Chapel Makefiles is to ensure that any rule encountered with `FORCE` as a prerequisite will always have its recipe executed, regardless of filesystem state.

However in failing to declare `FORCE` with a `.PHONY` directive, it means that if a user happens to have a file called `"FORCE"` in any directory containing such a Makefile (ie in the Chapel source tree), that Makefile might no longer operate as intended (seeing the on-disk file as confirmation that `FORCE` is up-to-date and then skip remaking any dependent rules). So for example in this case the `clean`, `cleanall` and `clobber` Makefile targets would become no-ops in a directory containing a stray file called `"FORCE"`.

It's admittedly a perverse/unusual use case, but it's also exactly what `.PHONY` is designed to concisely mitigate and IMHO it's best-practice for such targets in Makefiles.

Comment (bradcray) [reply third-party/gasnet/Makefile:162]: Got it, and thanks for the lesson.  Makes complete sense.

Comment (bradcray) [reply third-party/gasnet/Makefile:162]: I've opened https://github.com/chapel-lang/chapel/issues/26548 for this.

File: runtime/src/tasks/qthreads/tasks-qthreads.c (hunk 3 above)
Comment (bradcray) [reply runtime/src/tasks/qthreads/tasks-qthreads.c]: Improved this in https://github.com/chapel-lang/chapel/pull/26448/commits/d44edd2a8559f719ffa5789260a4f147bc6d66ee and caught/fixed a longstanding bug in testing it.

**Review by jabraham17 — APPROVED** [review]

File: runtime/src/comm/gasnet/comm-gasnet-ex.c [hunk 5]
```diff
@@ -714,7 +714,15 @@ int chpl_comm_addr_gettable(c_nodeid_t node, void* start, size_t len)
 
 
 int32_t chpl_comm_getMaxThreads(void) {
-  return GASNETI_MAX_THREADS-1;
+  uint64_t maxGasnetThreads = gex_System_QueryMaxThreads();
```
Comment (bonachea) [inline runtime/src/comm/gasnet/comm-gasnet-ex.c:717]: @bradcray Thanks for noticing and fixing this closely related defect! 

For the record: the previous code was (mostly) correct for the default GASNet configuration where the soft and hard client thread limits are equally 256 (it was also the only option back in GASNet-1 days when this code was initially written).  Now that this PR raises the hard limit at configure time, the soft and hard limits are no longer equal so the difference matters, and this GASNet-EX query returns this process's soft limit that is in force at runtime, which is what we actually need this function to return.

One minor quibble: 
IIUC The prior code used `GASNETI_MAX_THREADS-1` because the `-1` was reserving one thread (the primordial one that runs `chpl_comm_init()`) for use by the Chapel runtime, leaving the rest available for creation by the tasking layer (the consumer of the value we are returning here). I suspect we still need to reserve that one thread to avoid an off-by-one error in our accounting of total threads that might eventually invoke a GASNet call.

TL;DR:

```suggestion
  uint64_t maxGasnetThreads = gex_System_QueryMaxThreads() - 1;
```

**Review by bonachea — CHANGES_REQUESTED** [review]
Noted one minor potential defect, otherwise LGTM!

File: runtime/src/comm/gasnet/comm-gasnet-ex.c (hunk 5 above)
Comment (bonachea) [reply runtime/src/comm/gasnet/comm-gasnet-ex.c:717]: This is resolved (in a slightly different manner) by 465e173

**Review by bonachea — APPROVED** [review]
LGTM!
